### ISSUE TITLE
Variable features and pairpos subtables

### ISSUE THREAD

**Issue by simoncozens** [fonttools/fonttools#2772]
This feature code doesn't do what it looks like it should:

```
lookup kern_Common {
    pos [p] g -5;
    pos [p] y (wght=900:-100 wght=400:-100);
} kern_Common;
```

This creates two class-based pairpos subtables (one for each value record format) and the second one never fires for the string "py".

Is this a bug? I think so. And I think a possible fix would be to promote the value record format of the first pair so that they can be placed in the same subtable. As I understand it, this is only a problem related to PairPos subtables, so I'd put the fix into `otlLib.builder.PairPosBuilder`.

**Comment by simoncozens** [comment]
Weirder and weirder. In `PairPosBuilder`, we create a separate builder for each value format:

https://github.com/fonttools/fonttools/blob/d5a08b3b96f405b0fdca42e3f1f6baa32c59c687/Lib/fontTools/otlLib/builder.py#L1401-L1405

This is a `ClassPairPosSubtableBuilder`, which does its work by calling `buildPairPosClassesSubtable`. But `buildPairPosClassesSubtable` explicitly handles heterogenous value record formats! It first finds the "highest" value format for the list of value records:

https://github.com/fonttools/fonttools/blob/d5a08b3b96f405b0fdca42e3f1f6baa32c59c687/Lib/fontTools/otlLib/builder.py#L2128-L2129

and then promotes all VRs to that format:

https://github.com/fonttools/fonttools/blob/d5a08b3b96f405b0fdca42e3f1f6baa32c59c687/Lib/fontTools/otlLib/builder.py#L2143-L2152

So we don't actually need to create separate builders, and we can simplify the `PairPosBuilder` code to this:

```python
        builder = ClassPairPosSubtableBuilder(self)
        for glyphclass1, value1, glyphclass2, value2 in self.pairs:
            if glyphclass1 is self.SUBTABLE_BREAK_:
                builder.addSubtableBreak()
                continue
            builder.addPair(glyphclass1, value1, glyphclass2, value2)
        subtables = []
        if self.glyphPairs:
            subtables.extend(buildPairPosGlyphs(self.glyphPairs, self.glyphMap))
        subtables.extend(builder.subtables())
        lookup = self.buildLookup_(subtables)
```

All tests pass and my feature code now does what I expect it to.

**Comment by anthrotype** [comment]
I'm trying to understand why it was done like that to begin with, and what the implications of your proposed change in #2776 are.
As far as I understand, there's two main reason one needs to break a PairPos lookup into multiple subtables (all format2 class based in this particular case): 1) all pairs in a subtable share the same value format; 2) classes within a subtable cannot/should not partially overlap with one another.
In your example _both_ of these circumstances occur: there are overlapping classes (`[p]` appears in both pos rules) and there are different value formats (one is static, the second is variable).
In the current code, the responsibility of breaking subtables because of the two above mentioned reasons is in two different places: on the one hand, at first, additional ClassPairPosSubtableBuilder objects are created when new value formats are seen; on the other hand, the `ClassPairPosSubtableBuilder.flush_` is called each time classes overlap, and this calls the `buildPairPosClassesSubtable` function to create a single subtable and append to the builder's subtables list as new pairs get added; finally the builders are sorted by their value format pairs (used as the key in a dict) to create the final list of subtables (which also means the variable value formats and respective subtables get sorted _after_ the ones with non-variable value format).

You suggested change basically removes the first-level splitting that occurs when sorting pairs into buckets each with same pair of value formats, and only use one ClassPairPosSubtableBuilder that will upgrade the value formats of the pairs with non-partially-overlapping classes to the most general one.

Does that make sense?
Having said that, I'm still not sure we should proceed with your suggested change, need more time to think about it.

**Comment by anthrotype** [comment]
> In your example both of these circumstances occur: there are overlapping classes ([p] appears in both pos rules) and there are different value formats (one is static, the second is variable).

correction: only the second condition (differing value formats) applies in the above example; the classes overlap perfectly (`[p]` in both pos rules) so they can be in the same subtable.

**Comment by simoncozens** [comment]
Right, exactly. They would go in the same subtable were it not for the difference in value record formats. However, we can (and should) overcome that difference by promoting the value records to the same format, and indeed this is something we have code to do - but currently we deliberately avoid using it and instead split the builders based on value record format!

**Comment by anthrotype** [comment]
two questions:

1) is it ever desirable to instead split the subtable with differing value formats instead merging them into one?
2) why is it a problem in your case that the subtable gets split (which would be overcome by keeping them into the same subtable)?

**Comment by anthrotype** [comment]
> why is it a problem in your case that the subtable gets split

the problem is that the rule in the second subtable with the variable pos is never evaluated because the first one matches (for the first glyph) and shaper stops there.

I wonder if we could find a way to do the value format merging only when required such as in this case, but otherwise keep the subtables split by differing value formats when it doesn't cause issues.

**Comment by simoncozens** [comment]
> I wonder if we could find a way to do the value format merging only when required such as in this case, but otherwise keep the subtables split by differing value formats when it doesn't cause issues.

I'm not sure it's worth the work. If you have any variable kerns in your font, then they are probably in the majority and the number of non-variable fonts would be very small. So you probably end up with smaller binary size by keeping them in one subtable rather than splitting to a second.

**Comment by behdad** [comment]
> Weirder and weirder. In `PairPosBuilder`, we create a separate builder for each value format:

Correct. That's a bug. I see if I can fix it.

**Comment by behdad** [comment]
That's not the only bug near that code. `ClassPairPosSubtableBuilder` also adds new subtables if the classes are not "mergeable". That logic is also b0rked.

**Comment by anthrotype** [comment]
@behdad can you elaborate _why_ it's a bug to create separate subtables for each value format? I understand that it would fix the cases highlighted by @simoncozens (where a pos rule in a subsequent subtable becomes unreachable a preceding subtable already covers it) and we should fix these; but would it not also create larger subtables with more "sparse" value records more generally?

**Comment by behdad** [comment]
> @behdad can you elaborate _why_ it's a bug to create separate subtables for each value format? I understand that it would fix the cases highlighted by @simoncozens (where a pos rule in a subsequent subtable becomes unreachable a preceding subtable already covers it) and we should fix these; but would it not also create larger subtables with more "sparse" value records more generally?

It's not a bug per se to create separate subtables for each value format. The bug is to put the same glyph in multiple subtables, because only one of those will be reachable. Unless the code has logic to never do that, creating multiple subtables would be wrong. Since correctness should be preferred over optimization, it would be wrong to create separate subtables in this case.

**Comment by anthrotype** [comment]
@behdad thanks for the clarification. Makes sense.
